### PR TITLE
Improve thread partitioning for 2nd degree relations

### PR DESCRIPTION
**Pull request by gsouquet**
A bundle of fixes for event partitioning for threads, to review alongside https://github.com/matrix-org/matrix-js-sdk/pull/2165  which holds the main logic

---
This PR currently has no changelog labels, so will not be included in changelogs.

Add one of: `T-Deprecation`, `T-Enhancement`, `T-Defect`, `T-Task` to indicate what type of change this is plus `X-Breaking-Change` if it's a breaking change.

Preview: https://pr7748--matrix-react-sdk.netlify.app
⚠️ Do you trust the author of this PR? Maybe this build will steal your keys or give you malware. Exercise caution. Use test accounts.

### CODE REVIEW

**Comment by t3chguy** [comment]
PR title no longer matches what it does

**Review by t3chguy — CHANGES_REQUESTED** [review]
See above

File: src/components/structures/MessagePanel.tsx [hunk 1]
```diff
@@ -467,6 +467,17 @@ export default class MessagePanel extends React.Component<IProps, IState> {
 
     // TODO: Implement granular (per-room) hide options
     public shouldShowEvent(mxEv: MatrixEvent, forceHideEvents = false): boolean {
```
Comment (andybalaam) [inline src/components/structures/MessagePanel.tsx:469]: Would be good to rename this, because it sounds like it's the opposite of shouldHideEvent, but it's not.

File: src/components/structures/MessagePanel.tsx [hunk 2]
```diff
@@ -467,6 +467,17 @@ export default class MessagePanel extends React.Component<IProps, IState> {
...
+        if (this.props.hideThreadedMessages
+            && SettingsStore.getValue("feature_thread")) {
```
Comment (andybalaam) [inline src/components/structures/MessagePanel.tsx:471]: Could fit on a single line, or if not I'd prefer `) {` on a new line, so it's easy to see where the block starts.

File: src/components/structures/MessagePanel.tsx [hunk 3]
```diff
@@ -482,17 +493,23 @@ export default class MessagePanel extends React.Component<IProps, IState> {
...
-        if (mxEv.isThreadRelation &&
-            !mxEv.isThreadRoot &&
-            this.props.hideThreadedMessages &&
-            SettingsStore.getValue("feature_thread")
-        ) {
+        return !shouldHideEvent(mxEv, this.context);
+    }
+
+    private shouldLiveInThreadOnly(event: MatrixEvent): boolean {
+        const parentEventId = event.getAssociatedId();
```
Comment (andybalaam) [inline src/components/structures/MessagePanel.tsx:500]: Rename to associatedEventId?

File: src/components/structures/MessagePanel.tsx [hunk 4]
```diff
@@ -482,17 +493,23 @@ export default class MessagePanel extends React.Component<IProps, IState> {
...
+
+        const targetsThreadRoot = event.threadRootId === parentEventId;
+        if (event.isThreadRoot || targetsThreadRoot || !event.isThreadRelation) {
             return false;
         }
 
-        return !shouldHideEvent(mxEv, this.context);
+        const parentEvent = this.props.room.findEventById(parentEventId);
+        if (parentEvent) {
```
Comment (andybalaam) [inline src/components/structures/MessagePanel.tsx:508]: Might be worth adding a comment like:

```
// If this is a reply, then we use the thing we're replying to to decide whether we are displayed
```

Comment (andybalaam) [reply src/components/structures/MessagePanel.tsx:508]: (Or better wording ;-) )

File: src/components/structures/TimelinePanel.tsx [hunk 5]
```diff
@@ -1283,11 +1283,31 @@ class TimelinePanel extends React.Component<IProps, IState> {
...
+                    if (hasNoRelation) {
+                        return true;
+                    }
+
+                    if (e.isThreadRelation) {
+                        return false;
+                    }
+
+                    const parentEvent = this.props.timelineSet
+                        .findEventById(e.getAssociatedId());
```
Comment (andybalaam) [inline src/components/structures/TimelinePanel.tsx:1303]: Could all fit on one line?

File: src/components/views/rooms/EventTile.tsx [hunk 6]
```diff
@@ -391,6 +391,8 @@ export default class EventTile extends React.Component<IProps, IState> {
         this.context = context;
         const thread = this.thread;
 
+        this.onReactionsCreated = this.onReactionsCreated.bind(this);
```
Comment (andybalaam) [inline src/components/views/rooms/EventTile.tsx:394]: Why the change to a function here?  Was this an experiment around the event listener we looked at, that is no longer needed?

**Review by andybalaam — CHANGES_REQUESTED** [review]

**Review by andybalaam — APPROVED** [review]
Looks good apart from minor changes.

File: src/components/structures/MessagePanel.tsx (hunk 1 above)
Comment (gsouquet) [reply src/components/structures/MessagePanel.tsx:469]: Agree with you, but will probably leave that one out of this PR as I don't want to disturb that area of the codebase more than I currently have
